src/NHibernate/Type/CollectionType.cs

```diff
@@ -391,14 +392,36 @@ public virtual object InstantiateResult(object original)
 			// in which case we should never ever ever reset the dirty flag
 			// on the target because we simply do not know...
 			IPersistentCollection originalPc = original as IPersistentCollection;
-			IPersistentCollection resultPc = result as IPersistentCollection;
-			if(originalPc != null && resultPc!=null)
+			IPersistentCollection resultPc = target as IPersistentCollection;
+			if(resultPc!=null)
 			{
-				if (!originalPc.IsDirty)
+				if (elementsEqual || (originalPc != null && !originalPc.IsDirty))
 					resultPc.ClearDirty();
 			}
 
-			return result;
+			return target;
+		}
+
+		protected virtual bool AreCollectionElementsEqual(IEnumerable original, IEnumerable target)
```

Comment:
This method ends up inherited by `GenericMapType`, `GenericBagType`, `GenericIdentifierBagType` too. As for the `GenericSetType`, they are not sensitive to order (if they contains the same elements, they should be considered equal even if the order is not the same).

Maps requires checking they have the same elements count, that each key of one of them are found in the other one with the same value associated. It should be doable as an O(n) algorithm.

Bags allow duplicates. Checking two bags are equal implies checking they contain the same set of elements whatever their order, and counting how many time each element appears in each bags and checking the counts are the same. O(n) complexity is probably still achievable by building two map of their elements to their duplication count and then comparing the two maps.

@hazzik, too much added complexity?

Should we rename this method `AreCollectionsUndoubtedlyEqual` instead and just return false for those other types?